tsdb/engine/bz1/bz1.go

```diff
@@ -141,104 +169,163 @@ func (e *Engine) LoadMetadataIndex(index *tsdb.DatabaseIndex, measurementFields
+	if err := e.WAL.LoadMetadataIndex(index, measurementFields); err != nil {
+		return err
+	}
+
+	// finally open the WAL up
+	return e.WAL.Open()
 }
 
 // WritePoints writes metadata and point data into the engine.
 // Returns an error if new points are added to an existing key.
 func (e *Engine) WritePoints(points []tsdb.Point, measurementFieldsToSave map[string]*tsdb.MeasurementFields, seriesToCreate []*tsdb.SeriesCreate) error {
-	// Write series & field metadata.
-	if err := e.db.Update(func(tx *bolt.Tx) error {
-		if err := e.writeSeries(tx, seriesToCreate); err != nil {
+	// Write points to the WAL.
+	if err := e.WAL.WritePoints(points, measurementFieldsToSave, seriesToCreate); err != nil {
+		return fmt.Errorf("write points: %s", err)
+	}
+
+	return nil
+}
+
+// WriteIndex writes marshaled points to the engine's underlying index.
+func (e *Engine) WriteIndex(pointsByKey map[string][][]byte, measurementFieldsToSave map[string]*tsdb.MeasurementFields, seriesToCreate []*tsdb.SeriesCreate) error {
+	return e.db.Update(func(tx *bolt.Tx) error {
+		// Write series & field metadata.
+		if err := e.writeNewSeries(tx, seriesToCreate); err != nil {
 			return fmt.Errorf("write series: %s", err)
 		}
-		if err := e.writeFields(tx, measurementFieldsToSave); err != nil {
+		if err := e.writeNewFields(tx, measurementFieldsToSave); err != nil {
 			return fmt.Errorf("write fields: %s", err)
 		}
 
+		for key, values := range pointsByKey {
+			if err := e.writeIndex(tx, key, values); err != nil {
+				return fmt.Errorf("write: key=%x, err=%s", key, err)
+			}
+		}
 		return nil
-	}); err != nil {
+	})
+}
+
+func (e *Engine) writeNewFields(tx *bolt.Tx, measurementFieldsToSave map[string]*tsdb.MeasurementFields) error {
+	if len(measurementFieldsToSave) == 0 {
+		return nil
+	}
+
+	// read in all the previously saved fields
+	fields, err := e.readFields(tx)
+	if err != nil {
 		return err
 	}
 
-	// Write points to the WAL.
-	if err := e.PointsWriter.WritePoints(points); err != nil {
-		return fmt.Errorf("write points: %s", err)
+	// add the new ones or overwrite old ones
+	for name, mf := range measurementFieldsToSave {
+		fields[name] = mf
 	}
 
-	return nil
+	return e.writeFields(tx, fields)
 }
 
-// writeSeries writes a list of series to the metadata.
-func (e *Engine) writeSeries(tx *bolt.Tx, a []*tsdb.SeriesCreate) error {
-	// Ignore if there are no series.
-	if len(a) == 0 {
-		return nil
+func (e *Engine) writeFields(tx *bolt.Tx, fields map[string]*tsdb.MeasurementFields) error {
+	// compress and save everything
+	data, err := json.Marshal(fields)
+	if err != nil {
+		return err
 	}
 
-	// Marshal and insert each series into the metadata.
-	b := tx.Bucket([]byte("series"))
-	for _, sc := range a {
-		// Marshal series into bytes.
-		data, err := sc.Series.MarshalBinary()
-		if err != nil {
-			return fmt.Errorf("marshal series: %s", err)
-		}
+	return tx.Bucket([]byte("meta")).Put([]byte("fields"), snappy.Encode(nil, data))
```

Comment:
Is meta data that large that compression was a win here? Relative to the data itself, isn't meta data rather small?